include/dali/core/geom/mat.h

```diff
@@ -539,6 +540,82 @@ std::ostream &operator<<(std::ostream &os, const dali::mat<rows, cols, T> &m) {
   return os;
 }
 
+template <int n, int m, typename T>
+DALI_HOST_DEV
+inline void solve_gauss(mat<n, n, T> &A, mat<n, m, T> &B) {
+  for (int v = 0; v < n; v++) {
+    T max = std::abs(A(v, v));
+    int maxr = v;
+    for (int i = v + 1; i < n; i++) {
+      T q = std::abs(A(i, v));
+      if (q > max) {
+        max = q;
+        maxr = i;
+      }
+    }
+    if (!max) {
+      #ifdef __CUDA_ARCH__
+        continue;
+      #else
+        throw std::range_error("Cannot solve the system with a singlular matrix.");
+      #endif
+    }
+    if (maxr != v) {
+      for (int j = 0; j < n; j++)
+        cuda_swap(A(v, j), A(maxr, j));
+      for (int j = 0; j < m; j++)
+        cuda_swap(B(v, j), B(maxr, j));
+    }
+    T x = A(v, v);
+    x = 1/x;
+    A(v, v) = 1;
+    for (int j = v + 1; j < n; j++)
+      A(v, j) *= x;
+    for (int j = 0; j < n; j++)
+      B(v, j) *= x;
+    for (int i = 0; i < n; i++) {
+       if (i == v)
+        continue;
+#ifndef __CUDA_ARCH__
```

Comment:
I think this would look better at the top of the file, but maybe it is just me

Reply:
If we do it on top level, it would be in scope in every file that includes mat.h. Not sure if it's ok.